File: WoofAndroid/presentation/account/src/main/kotlin/com/perrystreet/woof/presentation/account/viewmodel/AccountViewModel.kt
package com.perrystreet.woof.presentation.account.viewmodel

import com.perrystreet.woof.presentation.account.mapper.DogProfileDomainToAccountUIModelMapper
import com.perrystreet.woof.presentation.account.uimodel.AccountUIModel
import com.perrystreet.woof.presentation.common.viewmodel.StateProducingViewModel
import com.perrystreet.woof.usecase.account.GetAccountUseCase
import io.reactivex.rxjava3.kotlin.plusAssign
import org.koin.android.annotation.KoinViewModel
import java.util.Optional

@KoinViewModel
class AccountViewModel(
    private val getAccountUseCase: GetAccountUseCase,
    private val accountMapper: DogProfileDomainToAccountUIModelMapper,
) : StateProducingViewModel<AccountViewModel.State>(initialValue = State.Loading) {

    sealed class State {
        data object Loading : State()

        data class Loaded(val account: AccountUIModel) : State()
    }

    override fun onFirstAppear() {
        disposables += getAccountUseCase().subscribe(
            { profile -> _state.onNext(State.Loaded(accountMapper(profile))) },
            { error -> mutableError.onNext(Optional.of(error)) },
        )
    }
}

File: WoofIOS/SwiftPackages/Presentation/Account/Sources/PresentationAccount/ViewModel/AccountViewModel.swift
import Combine
import DI
import Models
import PresentationCommon
import UseCase
import Utils

@Factory
public final class AccountViewModel: StateProducingViewModel<AccountViewModel.State, AccountError> {
    private let getAccountUseCase: GetAccountUseCase
    private let accountMapper: DogProfileDomainToAccountUIModelMapper

    init(
        getAccountUseCase: GetAccountUseCase,
        accountMapper: DogProfileDomainToAccountUIModelMapper
    ) {
        self.getAccountUseCase = getAccountUseCase
        self.accountMapper = accountMapper
        super.init(initialValue: .loading)
    }

    public override func onFirstAppear() {
        getAccountUseCase()
            .pss_sink { [weak self] result in
                guard let self else { return }
                switch result {
                case .success(let profile): setState(.loaded(account: accountMapper(profile: profile)))
                case .failure(let error): emitError(error)
                }
            }
            .store(in: &cancellables)
    }
}

extension AccountViewModel {
    public enum State: Equatable {
        case loading
        case loaded(account: AccountUIModel)
    }
}
